File: frontend/src/lib/algorithms/dfs.ts
import { TransitGraph } from "@/lib/transit-graph";
import { LINE_LOOKUP } from "@/lib/transit-graph";
import { type RouteSearchResult, type PathEntry, makeSegment } from "./types";

export function dfsSearch(
  graph: TransitGraph,
  originId: string,
  destId: string,
  viaStations?: string[],
  maxTimeMin: number = 120.0,
): RouteSearchResult | null {
  if (graph.isNodeBlocked(originId) || graph.isNodeBlocked(destId)) return null;

  if (viaStations && viaStations.length > 0) {
    const waypoints = [originId, ...viaStations, destId];
    return multiWaypointDfs(graph, waypoints, maxTimeMin);
  }

  return singleDfs(graph, originId, destId, maxTimeMin);
}

function singleDfs(
  graph: TransitGraph,
  originId: string,
  destId: string,
  maxTimeMin: number,
): RouteSearchResult | null {
  let best: RouteSearchResult | null = null;

  type StackEntry = {
    current: string;
    path: PathEntry[];
    timeAcc: number;
    distAcc: number;
    visited: Set<string>;
  };

  const stack: StackEntry[] = [
    {
      current: originId,
      path: [[originId, ""]],
      timeAcc: 0,
      distAcc: 0,
      visited: new Set([originId]),
    },
  ];

  while (stack.length > 0) {
    const { current, path, timeAcc, distAcc, visited } = stack.pop()!;

    if (current === destId) {
      const candidate = buildResult(graph, path, timeAcc, distAcc);
      if (best === null || timeAcc < best.total_time_min) {
        best = candidate;
      }
      continue;
    }

    if (timeAcc > maxTimeMin) continue;

    for (const edge of graph.getTraversableNeighbors(current)) {
      if (!visited.has(edge.to_node_id)) {
        const effTime = graph.getEffectiveTravelTime(current, edge);
        const newTime = timeAcc + effTime;
        if (newTime <= maxTimeMin) {
          const newVisited = new Set(visited);
          newVisited.add(edge.to_node_id);
          stack.push({
            current: edge.to_node_id,
            path: [...path, [edge.to_node_id, edge.line_id]],
            timeAcc: newTime,
            distAcc: distAcc + edge.distance_km,
            visited: newVisited,
          });
        }
      }
    }
  }

  return best;
}

function multiWaypointDfs(
  graph: TransitGraph,
  waypoints: string[],
  maxTimeMin: number,
): RouteSearchResult | null {
  let fullPath: PathEntry[] = [];
  let totalTime = 0;
  let totalDist = 0;
  const globalVisited = new Set<string>();

  for (let i = 0; i < waypoints.length - 1; i++) {
    const sub = singleDfsVisited(
      graph,
      waypoints[i],
      waypoints[i + 1],
      maxTimeMin - totalTime,
      globalVisited,
    );
    if (!sub) return null;

    const subPath = fullPath.length > 0 ? sub.path.slice(1) : sub.path;
    fullPath = fullPath.concat(subPath);
    totalTime += sub.time;
    totalDist += sub.dist;
    for (const p of subPath) globalVisited.add(p[0]);
  }

  return buildResult(graph, fullPath, totalTime, totalDist);
}

function singleDfsVisited(
  graph: TransitGraph,
  originId: string,
  destId: string,
  maxTimeMin: number,
  alreadyVisited: Set<string>,
): { path: PathEntry[]; time: number; dist: number } | null {
  let best: { path: PathEntry[]; time: number; dist: number } | null = null;

  type StackEntry = {
    current: string;
    path: PathEntry[];
    timeAcc: number;
    distAcc: number;
    visited: Set<string>;
  };

  const initVisited = new Set(alreadyVisited);
  initVisited.add(originId);

  const stack: StackEntry[] = [
    {
      current: originId,
      path: [[originId, ""]],
      timeAcc: 0,
      distAcc: 0,
      visited: initVisited,
    },
  ];

  while (stack.length > 0) {
    const { current, path, timeAcc, distAcc, visited } = stack.pop()!;

    if (current === destId) {
      if (best === null || timeAcc < best.time) {
        best = { path, time: timeAcc, dist: distAcc };
      }
      continue;
    }

    if (timeAcc > maxTimeMin) continue;

    for (const edge of graph.getTraversableNeighbors(current)) {
      const nid = edge.to_node_id;
      if (!visited.has(nid) || nid === destId) {
        const effTime = graph.getEffectiveTravelTime(current, edge);
        const newTime = timeAcc + effTime;
        if (newTime <= maxTimeMin) {
          const newVisited = new Set(visited);
          newVisited.add(nid);
          stack.push({
            current: nid,
            path: [...path, [nid, edge.line_id]],
            timeAcc: newTime,
            distAcc: distAcc + edge.distance_km,
            visited: newVisited,
          });
        }
      }
    }
  }

  return best;
}

function buildResult(
  graph: TransitGraph,
  path: PathEntry[],
  totalTime: number,
  totalDist: number,
): RouteSearchResult {
  const segments = [];
  let segStart = 0;

  for (let i = 2; i < path.length; i++) {
    if (path[i][1] !== path[i - 1][1] && path[i - 1][1] !== "") {
      const lineId =
        path[i - 1][1] ||
        (segStart + 1 < path.length ? path[segStart + 1][1] : "WALK");
      segments.push(makeSegment(graph, path, segStart, i - 1, lineId));
      segStart = i - 1;
    }
  }

  if (segStart < path.length - 1) {
    const lineId =
      path[path.length - 1][1] ||
      (segStart + 1 < path.length ? path[segStart + 1][1] : "WALK");
    segments.push(makeSegment(graph, path, segStart, path.length - 1, lineId));
  }

  const totalCo2 = segments.reduce((s, seg) => s + seg.co2_grams, 0);
  let transfers = 0;
  for (let i = 1; i < segments.length; i++) {
    if (segments[i].line !== segments[i - 1].line) transfers++;
  }

  return {
    segments,
    total_time_min: Math.round(totalTime * 10) / 10,
    total_distance_km: Math.round(totalDist * 100) / 100,
    total_transfers: transfers,
    total_co2_grams: Math.round(totalCo2 * 10) / 10,
    visited_ids: path.map((p) => p[0]),
  };
}

File: backend/app/algorithms/dfs.py
"""
DFS — Sightseeing / Constrained Path (Disruption-Aware).

Generates a continuous path that visits a specific sequence of landmark nodes
(via_stations) without revisiting any node, bounded by max_time_min.

Uses iterative DFS with backtracking.  When via_stations is provided,
the algorithm plans sub-paths between consecutive waypoints.

Blocked nodes/edges are excluded via ``get_traversable_neighbors``.
Travel times include incident penalties via ``get_effective_travel_time``.
Each result segment carries a ``co2_grams`` field.
"""

from __future__ import annotations

from typing import Optional, TYPE_CHECKING

from app.data.chennai_network import LINES, LINE_LOOKUP

if TYPE_CHECKING:
    from app.data.chennai_network import TransitGraph


def dfs_search(
    graph: TransitGraph,
    origin_id: str,
    dest_id: str,
    via_stations: list[str] | None = None,
    max_time_min: float = 120.0,
) -> Optional[dict]:
    if graph.is_node_blocked(origin_id) or graph.is_node_blocked(dest_id):
        return None

    if via_stations:
        waypoints = [origin_id] + via_stations + [dest_id]
        return _multi_waypoint_dfs(graph, waypoints, max_time_min)

    return _single_dfs(graph, origin_id, dest_id, max_time_min)


def _single_dfs(
    graph: TransitGraph,
    origin_id: str,
    dest_id: str,
    max_time_min: float,
) -> Optional[dict]:
    """DFS from origin to dest with time budget, no revisits."""

    best: dict | None = None

    stack: list[tuple[str, list[tuple[str, str]], float, float, set[str]]] = []
    stack.append((origin_id, [(origin_id, "")], 0.0, 0.0, {origin_id}))

    while stack:
        current, path, time_acc, dist_acc, visited = stack.pop()

        if current == dest_id:
            candidate = _build_result(graph, path, time_acc, dist_acc)
            if best is None or time_acc < best["total_time_min"]:
                best = candidate
            continue

        if time_acc > max_time_min:
            continue

        for edge in graph.get_traversable_neighbors(current):
            if edge.to_node_id not in visited:
                eff_time = graph.get_effective_travel_time(current, edge)
                new_time = time_acc + eff_time
                if new_time <= max_time_min:
                    new_visited = visited | {edge.to_node_id}
                    stack.append((
                        edge.to_node_id,
                        path + [(edge.to_node_id, edge.line_id)],
                        new_time,
                        dist_acc + edge.distance_km,
                        new_visited,
                    ))

    return best


def _multi_waypoint_dfs(
    graph: TransitGraph,
    waypoints: list[str],
    max_time_min: float,
) -> Optional[dict]:
    """Chain DFS sub-paths through each consecutive pair of waypoints."""

    full_path: list[tuple[str, str]] = []
    total_time = 0.0
    total_dist = 0.0
    global_visited: set[str] = set()

    for i in range(len(waypoints) - 1):
        sub = _single_dfs_visited(
            graph,
            waypoints[i],
            waypoints[i + 1],
            max_time_min - total_time,
            global_visited,
        )
        if sub is None:
            return None

        if full_path:
            sub_path = sub["path"][1:]
        else:
            sub_path = sub["path"]

        full_path.extend(sub_path)
        total_time += sub["time"]
        total_dist += sub["dist"]
        global_visited.update(p[0] for p in sub_path)

    return _build_result(graph, full_path, total_time, total_dist)


def _single_dfs_visited(
    graph: TransitGraph,
    origin_id: str,
    dest_id: str,
    max_time_min: float,
    already_visited: set[str],
) -> Optional[dict]:
    """DFS respecting already-visited nodes (except origin and dest)."""

    best: dict | None = None

    stack: list[tuple[str, list[tuple[str, str]], float, float, set[str]]] = []
    init_visited = already_visited.copy()
    init_visited.add(origin_id)

    stack.append((origin_id, [(origin_id, "")], 0.0, 0.0, init_visited))

    while stack:
        current, path, time_acc, dist_acc, visited = stack.pop()

        if current == dest_id:
            if best is None or time_acc < best["time"]:
                best = {"path": path, "time": time_acc, "dist": dist_acc}
            continue

        if time_acc > max_time_min:
            continue

        for edge in graph.get_traversable_neighbors(current):
            nid = edge.to_node_id
            if nid not in visited or nid == dest_id:
                eff_time = graph.get_effective_travel_time(current, edge)
                new_time = time_acc + eff_time
                if new_time <= max_time_min:
                    new_visited = visited | {nid}
                    stack.append((
                        nid,
                        path + [(nid, edge.line_id)],
                        new_time,
                        dist_acc + edge.distance_km,
                        new_visited,
                    ))

    return best


def _build_result(
    graph: TransitGraph,
    path: list[tuple[str, str]],
    total_time: float,
    total_dist: float,
) -> dict:
    segments: list[dict] = []
    seg_start = 0

    for i in range(2, len(path)):
        if path[i][1] != path[i - 1][1] and path[i - 1][1] != "":
            segments.append(_make_segment(graph, path, seg_start, i - 1))
            seg_start = i - 1

    if seg_start < len(path) - 1:
        segments.append(_make_segment(graph, path, seg_start, len(path) - 1))

    total_co2 = sum(s["co2_grams"] for s in segments)
    transfers = sum(
        1
        for i in range(1, len(segments))
        if segments[i]["line"] != segments[i - 1]["line"]
    )

    return {
        "segments": segments,
        "total_time_min": round(total_time, 1),
        "total_distance_km": round(total_dist, 2),
        "total_transfers": transfers,
        "total_co2_grams": round(total_co2, 1),
        "visited_ids": [p[0] for p in path],
    }


def _make_segment(
    graph: TransitGraph,
    path: list[tuple[str, str]],
    start: int,
    end: int,
) -> dict:
    from_s = graph.get_station(path[start][0])
    to_s = graph.get_station(path[end][0])

    line_id = (
        path[end][1]
        if path[end][1]
        else (path[start + 1][1] if start + 1 < len(path) else "WALK")
    )
    line_info = LINE_LOOKUP.get(line_id)
    mode = line_info.mode if line_info else "metro"

    total_time = 0.0
    total_dist = 0.0
    total_co2 = 0.0
    intermediates: list = []

    for i in range(start, end):
        sid = path[i][0]
        nid = path[i + 1][0]
        found = False
        for edge in graph.get_neighbors(sid):
            if edge.to_node_id == nid and edge.line_id == line_id:
                total_time += graph.get_effective_travel_time(sid, edge)
                total_dist += edge.distance_km
                total_co2 += edge.emission_gco2_per_km * edge.distance_km
                found = True
                break
        if not found:
            for edge in graph.get_neighbors(sid):
                if edge.to_node_id == nid:
                    total_time += graph.get_effective_travel_time(sid, edge)
                    total_dist += edge.distance_km
                    total_co2 += edge.emission_gco2_per_km * edge.distance_km
                    break

        if start < i < end:
            intermediates.append(graph.get_station(path[i][0]))

    return {
        "from_station": from_s,
        "to_station": to_s,
        "line": line_id,
        "mode": mode,
        "travel_time_min": round(total_time, 1),
        "distance_km": round(total_dist, 2),
        "co2_grams": round(total_co2, 1),
        "intermediate_stations": intermediates,
    }
